MPTCP: implementation of mptcp

mux99 opened the pull request:
This PR is internal, and aims to implement MPTCP (defined in RFC8684) into the nginx socket creation process.
The current approach is to add an option to the `listen` directive (namely `mptcp`) to enable and disable the protocol.

It would be interesting to have it set by the `http`or `server` blocks but I don't currently see a good way of doing so. 

In this initial commit MPTCP is disabled by default simply because it is easy to implement, but it is preferable to have is on by default and give the option to user to explicitly disable it.

Commit:
MPTCP: test implementation

this is a test implementatation of mptcp fot the http context of the nginx server

matttbe commented on auto/unix:
```diff
@@ -533,6 +533,15 @@ ngx_feature_libs=
+ngx_feature_path=
+ngx_feature_libs=
+ngx_feature_test="socket(AF_INET, SOCK_STREAM, IPPROTO_MPTCP)"
+. auto/feature
```
best to avoid doing that at build time. Save that on a branch, just in case the maintainers insist doing that, but best to avoid.

mux99 replied:
done

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -487,8 +488,12 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
+                            ls[i].protocol);
+#else
             s = ngx_socket(ls[i].sockaddr->sa_family, ls[i].type, 0);
-
```
:-1:

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -487,8 +488,12 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
 
+#if (NGX_HAVE_MPTCP)
+            s = ngx_socket(ls[i].sockaddr->sa_family, ls[i].type,
+                            ls[i].protocol);
```
alignment not ok

mux99 replied:
is it ok now? not quite sure

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -487,8 +488,12 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
                 continue;
             }
 
+#if (NGX_HAVE_MPTCP)
```
strange to do that here, you don't know what `protocol` is

mux99 replied:
this is where the default value of `protocol`is set, to keep the default behavior of nginx i put it to 0

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -487,8 +488,12 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
             }
 
+#if (NGX_HAVE_MPTCP)
+            s = ngx_socket(ls[i].sockaddr->sa_family, ls[i].type,
```
we should retry without the protocol if there was an error

  s = (ngx_socket_t) -1;
  if (protocol > 0)
      s = (...)
  if (s == (ngx_socket_t) -1)
      s = (...)

mux99 replied:
done

Commit:
fix: replace socket call swap w/ fallback

Commit:
fix: alignment, hopefully

Commit:
fix: removed compilator mptcp test

Commit:
add: an option mptcp for streams

matttbe commented on auto/unix:
```diff
@@ -533,7 +533,6 @@ ngx_feature_libs=
 ngx_feature_test="setsockopt(0, IPPROTO_TCP, TCP_FASTOPEN, NULL, 0)"
 . auto/feature
 
-
```
:-1: (remaining modified line)

Please always your modifications before sending them, these unwanted modifications are easy to spot.

(also, you can amend/squash your commits)

mux99 replied:
done

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -487,6 +488,13 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
                 continue;
             }
 
+#if (NGX_LINUX)
```
if we can can do that on other OS, then you don't need this check here, just making sure `protocol` is initialised to 0 (I think it is)

(e.g. `lsopt` used below has been explicitly `ngx_memzero()`)

mux99 replied:
in the same block, would it be considered to expensive to always call socket twice if it failed the first time?

matttbe replied:
The idea would be that the protocol is > 0 only in Linux.

mux99 replied:
ok, done then

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -78,6 +78,7 @@ ngx_create_listening(ngx_conf_t *cf, struct sockaddr *sockaddr,
 
     ls->fd = (ngx_socket_t) -1;
     ls->type = SOCK_STREAM;
+    ls->protocol = 0;
```
check how `ls` is initialised: maybe they do a calloc, and they only have to set non 0 values.

mux99 replied:
correct, there is a memzero() call for ls. removing the line ^^

matttbe commented on src/core/ngx_connection.c:
```diff
@@ -487,6 +488,13 @@ ngx_open_listening_sockets(ngx_cycle_t *cycle)
+                               ls[i].protocol);
+            if (s == (ngx_socket_t) -1)
+#endif
             s = ngx_socket(ls[i].sockaddr->sa_family, ls[i].type, 0);
```
this needs to be indented now, to go with the 'if' above.

Commit:
fix: spacing

Commit:
fix: removed useless pre-processors

Commit:
fix: value already zero

Commit:
MPTCP: test implementation

this is a test implementatation of mptcp fot the http context of the nginx server

fix: replace socket call swap w/ fallback

fix: alignment, hopefully

fix: removed compilator mptcp test

add: an option mptcp for streams

fix: spacing

fix: removed useless pre-processors

fix: value already zero

Commit:
Merge branch 'mptcp' of https://github.com/mux99/nginx into mptcp

mux99 commented:
my bad, branch is back

matttbe commented:
> my bad, branch is back

I guess you need to re-open the PR, or create a new one.

mux99 commented:
still not proficient with git. i created a new PR
